Source/Diagnostics/ReducedDiags/LoadBalanceCosts.cpp

```diff
@@ -81,20 +86,96 @@ void LoadBalanceCosts::ComputeDiags (int step)
         for (MFIter mfi(Ex, false); mfi.isValid(); ++mfi)
         {
             const Box& tbx = mfi.tilebox();
-            m_data[shift + mfi.index()*m_nDataFields + 0] = (*costs[lev])[mfi.index()];
-            m_data[shift + mfi.index()*m_nDataFields + 1] = dm[mfi.index()];
-            m_data[shift + mfi.index()*m_nDataFields + 2] = lev;
-            m_data[shift + mfi.index()*m_nDataFields + 3] = tbx.loVect()[0];
-            m_data[shift + mfi.index()*m_nDataFields + 4] = tbx.loVect()[1];
-            m_data[shift + mfi.index()*m_nDataFields + 5] = tbx.loVect()[2];
+            m_data[shift_m_data + mfi.index()*m_nDataFields + 0] = (*costs[lev])[mfi.index()];
+            m_data[shift_m_data + mfi.index()*m_nDataFields + 1] = dm[mfi.index()];
+            m_data[shift_m_data + mfi.index()*m_nDataFields + 2] = lev;
+            m_data[shift_m_data + mfi.index()*m_nDataFields + 3] = tbx.loVect()[0];
+            m_data[shift_m_data + mfi.index()*m_nDataFields + 4] = tbx.loVect()[1];
+            m_data[shift_m_data + mfi.index()*m_nDataFields + 5] = tbx.loVect()[2];
         }
 
         // we looped through all the boxes on level lev, update the shift index
-        shift += m_nDataFields*(costs[lev]->size());
+        shift_m_data += m_nDataFields*(costs[lev]->size());
     }
 
     // parallel reduce to IO proc and get data over all procs
-    ParallelDescriptor::ReduceRealSum(m_data.data(), m_data.size(), ParallelDescriptor::IOProcessorNumber());
+    ParallelDescriptor::ReduceRealSum(m_data.data(),
+                                      m_data.size(),
+                                      ParallelDescriptor::IOProcessorNumber());
+
+#ifdef AMREX_USE_MPI
+    // now parallel reduce to IO proc and get string data (host name) over all procs
+    // MPI Gatherv preliminaries
+
+    // get the MPI host name and number of characters
+    char hostname[MPI_MAX_PROCESSOR_NAME];
+    int length;
+
+    MPI_CHECK( MPI_Get_processor_name( hostname, &length ) );
+
+    // IO proc will collect messages from other procs;
+    // receive counts and displacements needed only by IO proc
+    if (ParallelDescriptor::MyProc() == ParallelDescriptor::IOProcessorNumber())
+    {
+        m_data_string_recvcount.resize(ParallelDescriptor::NProcs(), 0);
+        m_data_string_recvcount.assign(ParallelDescriptor::NProcs(), 0);
+        m_data_string_disp.resize(ParallelDescriptor::NProcs(), 0);
+        m_data_string_disp.assign(ParallelDescriptor::NProcs(), 0);
+    }
+
+    // get the string lengths on IO proc
+    ParallelDescriptor::Gather(&length, 1,                     // send
+                               &m_data_string_recvcount[0], 1, // receive
+                               ParallelDescriptor::IOProcessorNumber());
+
+    // determine total length of collected strings for root, and set displacements
+    if (ParallelDescriptor::MyProc() == ParallelDescriptor::IOProcessorNumber())
+    {
+        // + 1 is for separator character
```

Comment:
If user provides a special separator, the length of it could be greater than 1. For example, a double whitespace rather than a single whitespace. It would be better to figure out the length of the separator from `m_sep`?

Reply:
good point, the comment I put there is actually not clear (for the updated version I have clarified).  the separator here is not the same as `m_sep`, but just a chosen delimiter to separate words in the gathered string, which is later split (see `m_data_string = WarpXUtilStr::splitString(m_data_string_recvbuf, ' ');` further down).  so this separator is hidden from the user, but chosen by hand as `' '`.